pkg/ingress/v1/istio/client.go

```diff
@@ -0,0 +1,214 @@
+package istio
+
+import (
+	"fmt"
+	"maps"
+	"strconv"
+
+	istioclient "istio.io/client-go/pkg/clientset/versioned"
+	isnetlister "istio.io/client-go/pkg/listers/networking/v1beta1"
+	isseclister "istio.io/client-go/pkg/listers/security/v1beta1"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	netv1beta1 "istio.io/api/networking/v1beta1"
+	secv1beta1 "istio.io/api/security/v1beta1"
+	isv1beta1 "istio.io/api/type/v1beta1"
+	isnetv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	issecv1beta1 "istio.io/client-go/pkg/apis/security/v1beta1"
+
+	common "github.com/kanopy-platform/argoslower/pkg/ingress"
+)
+
+// IstioClient provides a means of upserting rendered resource from an IstioConfig object
+type IstioClient struct {
+	client istioclient.Interface
+	vsl    isnetlister.VirtualServiceLister
+	apl    isseclister.AuthorizationPolicyLister
+}
+
+func NewClient(cs istioclient.Interface, vsl isnetlister.VirtualServiceLister, apl isseclister.AuthorizationPolicyLister) *IstioClient {
+	return &IstioClient{
+		client: cs,
+		vsl:    vsl,
+		apl:    apl,
+	}
+}
+
+// UpsertFromConfig creates or updates Virtual Serivces associated with an IstioConfig
+// It returns an error for unconfigured IstioConfigs
+func (i *IstioClient) UpsertFromConfig(config *IstioConfig) error {
+
+	//TODO: code to create or apply VirtualService and AuthorizationPolicies
+
+	return nil
+}
+
+// IstioConfig contains global configuration for rendering istio VirtualService and AuthorizationPolicy
+// resources from a port to endpoint mapping.
+type IstioConfig struct {
+	ap         *issecv1beta1.AuthorizationPolicy
+	vs         *isnetv1beta1.VirtualService
+	gateway    types.NamespacedName
+	gwSelector map[string]string
+	baseURL    string
+}
+
+func NewIstioConfig(gw types.NamespacedName, gws map[string]string, baseURL string) *IstioConfig {
+	selector := maps.Clone(gws)
+	return &IstioConfig{
+		gateway:    gw,
+		gwSelector: selector,
+		baseURL:    baseURL,
+	}
+}
+
+// GetVirtualService returns the current configured or unconfigured virtual service
+// if the virtual service is unconfigured it is nil.
+func (ic *IstioConfig) GetVirtualService() *isnetv1beta1.VirtualService {
+	return ic.vs
```

Comment:
This should probably return a `DeepCopy` or the `vs` field could be public in the struct.

Reply:
Returning a deep copy is probably the right answer. The vs and ap fields are intentionally private. Outside of this package, the only way to configure the vs or ap is authoritatively the Configure* functions.